src/analysis/natural_parameterization.lean

```diff
@@ -0,0 +1,227 @@
+
+lemma is_linearly_parameterized_on_by_of_subsingleton
+  (f : ℝ → E) {s : set ℝ} (hs : s.subsingleton) (l : ℝ≥0) : is_linearly_parameterized_on_by f s l :=
+begin
+  rintro x hx y hy, cases hs hx hy,
+  rw evariation_on.subsingleton f (λ y hy z hz, hs hy.1 hz.1 : (s ∩ Icc x x).subsingleton),
+  simp only [sub_self, mul_zero, ennreal.of_real_zero],
+end
+
+lemma is_linearly_parameterized_on_by.iff_ordered :
+  is_linearly_parameterized_on_by f s l ↔
+  ∀ ⦃x⦄ (hx : x ∈ s) ⦃y⦄ (hy : y ∈ s), (x ≤ y) →
+    evariation_on f (s ∩ Icc x y) = ennreal.of_real (l * (y - x)) :=
+begin
+  refine ⟨λ h x xs y ys xy, h xs ys, λ h x xs y ys, _⟩,
+  rcases le_total x y with xy|yx,
+  { exact h xs ys xy, },
+  { rw [evariation_on.subsingleton, ennreal.of_real_of_nonpos],
+    { exact mul_nonpos_of_nonneg_of_nonpos l.prop (sub_nonpos_of_le yx), },
+    { rintro z ⟨zs, xz, zy⟩ w ⟨ws, xw, wy⟩,
+      cases le_antisymm (zy.trans yx) xz,
+      cases le_antisymm (wy.trans yx) xw,
+      refl, }, },
+end
+
+lemma is_linearly_parameterized_on_by.iff_variation_on_from_to_eq :
+  is_linearly_parameterized_on_by f s l ↔ (has_locally_bounded_variation_on f s ∧
+  ∀ ⦃x⦄ (hx : x ∈ s) ⦃y⦄ (hy : y ∈ s), variation_on_from_to f s x y = l * (y - x)) :=
+begin
+  split,
+  { rintro h, refine ⟨h.has_locally_bounded_variation_on, λ x xs y ys, _⟩,
+    rw is_linearly_parameterized_on_by.iff_ordered at h,
+    rcases le_total x y with xy|yx,
+    { rw [variation_on_from_to.eq_of_le f s xy, h xs ys xy,
+          ennreal.to_real_of_real (mul_nonneg l.prop (sub_nonneg.mpr xy))], },
+    { rw [variation_on_from_to.eq_of_ge f s yx, h ys xs yx ,
+          ennreal.to_real_of_real  (mul_nonneg l.prop (sub_nonneg.mpr yx)),
+          mul_comm ↑l, mul_comm ↑l, ←neg_mul, neg_sub], }, },
+  { rw is_linearly_parameterized_on_by.iff_ordered,
+    rintro h x xs y ys xy,
+    rw [←h.2 xs ys, variation_on_from_to.eq_of_le f s xy,
+        ennreal.of_real_to_real (h.1 x y xs ys)], },
+end
+
+lemma is_linearly_parameterized_on_by_zero_iff :
+  is_linearly_parameterized_on_by f s 0 ↔ ∀ x y ∈ s, edist (f x) (f y) = 0 :=
+begin
+  dsimp [is_linearly_parameterized_on_by],
+  simp only [zero_mul, ennreal.of_real_zero, ←evariation_on.eq_zero_iff],
+  split,
+  { by_contra',
+    obtain ⟨h, hfs⟩ := this,
+    simp_rw evariation_on.eq_zero_iff at hfs h,
+    push_neg at hfs,
+    obtain ⟨x, xs, y, ys, hxy⟩ := hfs,
+    rcases le_total x y with xy|yx,
+    { exact hxy (h xs ys x ⟨xs, le_rfl, xy⟩ y ⟨ys, xy, le_rfl⟩), },
+    { rw edist_comm at hxy,
+      exact hxy (h ys xs y ⟨ys, le_rfl, yx⟩ x ⟨xs, yx, le_rfl⟩), }, },
+  { rintro h x xs y ys,
+    refine le_antisymm _ (zero_le'),
+    rw ←h,
+    exact evariation_on.mono f (inter_subset_left s (Icc x y)), },
+end
+
+lemma is_linearly_parameterized_on_by.ratio {t : set ℝ} {φ : ℝ → ℝ}
+  (φm : monotone_on φ s) (φst : s.maps_to φ t) (φst' : s.surj_on φ t)
+  {l' : ℝ≥0} (hl' : l' ≠ 0)
```

Comment:
I think the lemma is true without `hl'`, so you should try hard to remove it. (The philosophy is to make things easy for the users of your API, not for yourself :-)

Reply:
If `l'` is zero, then so is `l`. If `f` is any function to a degenerate *pseudo* emetric space `E` where all points are at distance zero, then we can choose whatever we want for `\phi` and setting `l l' = 0`, we don't necessarily have `\phi` constant. 
Am I missing something?

Reply:
Actually, we can get even simpler, if `f` is any constant function `\phi` can be chosen freely with still `l l' = 0`.